### DISCUSSION TITLE
Pass SSL certificates contents as string instead of path

### DISCUSSION THREAD

**Discussion by nicholas-alonzo** [sqlalchemy/sqlalchemy#9863] [Usage Questions] [answered]
Hi,

I'm working on a project that involves connecting to Postgres using SSL certificates. I'm using SQLAlchemy and psycopg2 to connect like so:

```
ssl_args = {
    'sslmode': 'verify-ca',
    'sslcert': '/path/to/ssl.crt',
    'sslkey': '/path/to/ssl.key',
    'sslrootcert': '/path/to/root.crt'
}

db_engine = sqlalchemy.create_engine('postgresql+psycopg2://user:password@host:port/dbname, connect_args=ssl_args)
```

I'm now wondering if there's a valid way to pass the contents of the SSL certificates as a string to connect. I tried this below, but SQLAlchemy throws an error that it's not a valid path.

```
ssl_args = {
    'sslmode': 'verify-ca',
    'sslcert': '---------BEGIN....',
    'sslkey': '---------BEGIN....',
    'sslrootcert': '---------BEGIN....',
}

db_engine = sqlalchemy.create_engine('postgresql+psycopg2://user:password@host:port/dbname, connect_args=ssl_args)
```

I asked the same [question](https://github.com/psycopg/psycopg2/discussions/1583) on the psycopg2 discussions and was rerouted here. Any thoughts or ideas would be appreciated, thank you

**Comment by zzzeek** [comment]
im not sure what @dvarrazzo is suggesting, we send params to the driver and that's it.  is there a particular parameter at https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS you'd like to use?

**Reply by nicholas-alonzo** [reply]
Hi there, thanks for a reply. Yes, it would be the following SSL parameters to connect to Postgres:
- sslcert
- sslkey
- sslrootcert

To give more background, I have a KeePass DB that contains the Postgres DB credentials and SSL certificates that can be read into Python. The SSL certificates come back as a string, but an error get's thrown when trying to use it because it's not a valid path even though it's the certificate content used to connect.

**Reply by zzzeek** [reply]
pass them right through , see the docs at https://docs.sqlalchemy.org/en/20/core/engines.html#custom-dbapi-args which are pretty in-depth

```py
from sqlalchemy import create_engine

e = create_engine(
    "postgresql+psycopg2://user:pass@host/dbname",
    connect_args={
        "sslmode": "verify-ca",
        "sslcert": "---------BEGIN....",
        "sslkey": "---------BEGIN....",
        "sslrootcert": "---------BEGIN....",
    },
)

```

**Reply by zzzeek** [reply]
which is what you did, but you refer to a mysterious message with no exception object or stack trace "SQLAlchemy throws an error that it's not a valid path.", that sounds like psycopg2.

basically we dont do anything here, if you can do it with psycopg2 I can show you how to pass through with SQLAlchemy.  if psycopg2 can't do it, then you cant do it with that driver.   SQLAlchemy is not really a factor here.

**Reply by nicholas-alonzo** [reply]
Hi @zzzeek, thanks again for your replies. I figured it might be psycopg2 related and that's why I asked there first. Here's pieces of the error that I receive. I thought it mentioned the file path not being valid, but to correct myself the output says "does not exist" in reference to the root certificate.

```
psycopg2.OperationalError: connection to server at "HOST", port 5432 failed: root certificate file "-----BEGIN CERTIFICATE----- ..." does not exist 
Either provide the file or change sslmode to disable server certificate verification. 
.
.
.
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "HOST", port 5432 failed: root certificate file "-----BEGIN CERTIFICATE----- ..." does not exist 
Either provide the file or change sslmode to disable server certificate verification. 
.
.
.
https://docs.sqlalchemy.org/en/20/errors.html#error-e3q8 
```

To get around this error, I'm currently using the tempfile module to recreate the certificates so I can reference an actual file path. It's not really ideal, but it works. I can bring it up again on the psycopg2 repo.

**Reply by zzzeek** [reply, marked as answer]
`psycopg2.OperationalError` - that's the psycopg2 driver's error

> I'm currently using the tempfile module to recreate the certificates so I can reference an actual file path. 

I think this is how those parameters work.  I dont think there is an option to pass the string directly but you'd want to email on the Postgresql users lists regarding their C api to get the details

**Reply by nicholas-alonzo** [reply]
I'll ask and see what they think - much appreciated for your time. I'll mark your last reply as the answer
